pkg/metrics/providers/splunk.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2020 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package providers
+
+import (
+	"cmp"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/signalfx/signalflow-client-go/signalflow"
+	"github.com/signalfx/signalflow-client-go/signalflow/messages"
+
+	flaggerv1 "github.com/fluxcd/flagger/pkg/apis/flagger/v1beta1"
+)
+
+// https://docs.datadoghq.com/api/
+const (
+	signalFxSignalFlowApiPath = "/v2/signalflow"
+	signalFxValidationPath    = "/v2/metric?limit=1"
+
+	signalFxTokenSecretKey = "sf_token_key"
+
+	signalFxTokenHeaderKey = "X-SF-Token"
+
+	signalFxFromDeltaMultiplierOnMetricInterval = 10
+)
+
+// SplunkProvider executes signalfx queries
+type SplunkProvider struct {
+	metricsQueryEndpoint  string
+	apiValidationEndpoint string
+
+	timeout   time.Duration
+	token     string
+	fromDelta int64
+}
+
+type splunkResponse struct {
+}
+
+// NewSplunkProvider takes a canary spec, a provider spec and the credentials map, and
+// returns a Splunk client ready to execute queries against the API
+func NewSplunkProvider(metricInterval string,
+	provider flaggerv1.MetricTemplateProvider,
+	credentials map[string][]byte) (*SplunkProvider, error) {
+
+	address := provider.Address
+	if address == "" {
+		return nil, fmt.Errorf("splunk endpoint is not set")
+	}
+
+	sp := SplunkProvider{
+		timeout:               5 * time.Second,
+		metricsQueryEndpoint:  strings.Replace(strings.Replace(address+signalFxSignalFlowApiPath, "http", "ws", 1), "api", "stream", 1),
```

Comment:
could you add some context here?

Reply:
The library used to handle the signalflow api uses websockets to connect, so the address is converted.
I've added a COMMENT.